### ISSUE TITLE
Merging Registry and Poll (again)

### ISSUE THREAD

**Issue by Thomasdezeeuw** [tokio-rs/mio#1025]
In v0.7 `Poll` on unix is just an file descriptor, on the master branch it's split into `Registry` and `Poll` both using `Arc` to share the file descriptor. In my use case I don't need this split as I only use `Poll` from a single location, making the `Arc` unneeded overhead. I propose to remove `Registry` and move back to the structure that was present in v0.7, i.e. a single `Poll`. Making all methods on `Poll` take a reference to self so that the user can use `Arc` and make the split if needed.

**Comment by carllerche** [comment]
I would say we should wait until we have the windows update done... or at least are clear on the direction.

**Comment by carllerche** [comment]
The main issue is that, merging it back, would require all OSes being able to guarantee that polling is sync...

**Comment by Thomasdezeeuw** [comment]
I don't know about Windows, but aren't C thing `Sync`? As C doesn't really have a concept of `Sync`? But I'm fine with waiting on Windows.

**Comment by carllerche** [comment]
Yes, the FD / C is sync... it just is a pretty strong guarantee. I'm not necessarily against merging it back, but we should have a good sense of the pros / cons per platform.

**Comment by Thomasdezeeuw** [comment]
We could force it to be `!Sync`, then wrapping it an `Arc` still works ([example on playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=dd0aaf49d9fd219df6eee78d0c287764))

**Comment by carllerche** [comment]
Then it isn’t send right? The arc handle that is.

**Comment by Thomasdezeeuw** [comment]
Making `Poll` `!Send` and `!Sync` still makes it work with `Arc`: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=29120553d1e69fc6f1551ce5a6fad1ed.

**Comment by carllerche** [comment]
but, we need to be able to register sockets across threads, we only want to poll from one. Maybe I'm missing something, but if `Poll` is used to register sockets and it is `!Send + !Sync`, then it would not be possible to register sockets across threads.

**Comment by Thomasdezeeuw** [comment]
@carllerche you're right I forget to check the requirements of the `Deref` implementation of `Arc`. Lets wait until the Windows rewrite then to see what the requirements should be.

**Comment by carllerche** [comment]
Sounds good. Ideally we could get some validation for redox / fuchsia as well.

**Comment by Thomasdezeeuw** [comment]
@carllerche after the Windows rewrite `windows::Selector` is still `Sync` and `Send` (and so are the Unix counterparts). How do want to validate this for Redox/Fuchsia? Otherwise I would like to move forward with this requiring `Poll` and (thus the `Selector`) to be `Sync` and `Send`, removing the `Arc` and `Registry`.

**Comment by carllerche** [comment]
Another question is how to handle concurrency. It doesn’t make a ton of sense to concurrently poll on Linux (unless the exclusive flag is set, not sure about kqueue). Also, on windows there is a hard limit on the max number of concurrent polls. 

One thought I had with having a Poll handle take &mut self is they it limits concurrent polling. Then, we could add a Poll::try_clone fn on Poll. This would succeed if the selector supports additional concurrency. So, on windows we could make that work to the upper bound. On Linux we could possibly only make it work if the exclusive flag is supported. 

Thoughts?

**Comment by Thomasdezeeuw** [comment]
@carllerche do you mean to make all methods on `Poll` to take`&mut self`, i.e. `Poll::poll` but also `Poll::register` and friends? Either way I'm ok with it since I only use `Poll` from a single thread and have mutable access already, however I'm not sure how other people use `Poll`.

Alternatively we could keep `Registry` but make it optional and not behind an `Arc`. On Unix it would `dup(2)` the file descriptor of the selector and on Windows `Selector` is already an `Arc<SelectorInner>` so it would use that. As far as I can tell concurrent registering is fine. So `Poll::poll` would need `&mut self`, `Poll::register` and `Registry::register` (and friends) would only need `&self`.

**Comment by carllerche** [comment]
I would rather not add duping for `Register`.

Registering needs to be `Sync` (at the very least for Tokio), so `register` must take `&self`.

If `Registry` and `Poll` are merged, all fns will need to take `&self` for ^^ including `poll`. This would mean that `poll` would also have to be `Sync`. As you pointed out, this is true today and it is *probably* a safe assumption that all OSes support concurrent polling. Worst case scenario is adding a lock.

The *problem* is how windows deals with concurrent polling. See [here](https://docs.microsoft.com/en-us/windows/win32/fileio/i-o-completion-ports#threads-and-concurrency). Basically, the behavior you want with concurrent polling on windows is you set `NumberOfConcurrentThreads` and then only that number of threads will ever poll. If we merge `Poll` and `Registry`, we will not have a way of modeling this.

**Comment by Thomasdezeeuw** [comment]
In the [CreateIoCompletionPort](https://docs.microsoft.com/en-us/windows/win32/fileio/createiocompletionport) docs under `NumberOfConcurrentThreads` (can't link directly) it says:

> If this parameter is zero, the system allows as many concurrently running threads as there are processors in the system.

Is that processor core/hardware thread or processor? If it's per core wouldn't that fix the problem? I suppose you could create multiple threads per core and poll concurrently on all of those, but most models use 1 poll per application or 1 poll per thread.

Also can't we set it a very high number, e.g. 4294967295. It seems unreasonable that many threads will poll concurrently. However I have know idea if that will work and how much, if any, resources that needs OS side.

**Comment by carllerche** [comment]
@Thomasdezeeuw my read (which we probably should verify) is not the # of concurrent polls, but the # of threads that can be associated with the IOCP port at the same time. Read the above section and it describes how a thread becomes associated / unassociated with an IOCP port. My guess is that, once a thread is associated with the IOCP port, it will block another thread even if it is not currently polling.

I also don't think we can just set a very high number as my guess is internal memory is allocated / reserved proportionally to the number.

Maybe @piscisaureus knows more than I do, either way we probably should verify this.

**Comment by carllerche** [comment]
Specifically:

> Although any number of threads can call GetQueuedCompletionStatus for a specified I/O completion port, when a specified thread calls GetQueuedCompletionStatus the first time, it becomes associated with the specified I/O completion port until one of three things occurs: The thread exits, specifies a different I/O completion port, or closes the I/O completion port. In other words, a single thread can be associated with, at most, one I/O completion port.

**Comment by Thomasdezeeuw** [comment]
@carllerche I've created a proof of concept for my idea that allows `Poll::poll` to take `&mut self`, but register `&self` and keeps `Registry` for use behind an `Arc` or similar. The branch is available here: https://github.com/Thomasdezeeuw/mio/tree/poll-no-arc, the commit with the API: https://github.com/Thomasdezeeuw/mio/commit/667419626b128778453867aae23839c57528c836 and a commit with a tests that shows it works with an `Arc`: https://github.com/Thomasdezeeuw/mio/commit/8e81d0fca92a7e20dec9f413c704f4dbba70526c. What do you think?

**Comment by piscisaureus** [comment]
> @Thomasdezeeuw my read (which we probably should verify) is not the # of concurrent polls, but the # of threads that can be associated with the IOCP port at the same time. Read the above section and it describes how a thread becomes associated / unassociated with an IOCP port. My guess is that, once a thread is associated with the IOCP port, it will block another thread even if it is not currently polling.
> 
> I also don't think we can just set a very high number as my guess is internal memory is allocated / reserved proportionally to the number.
> 
> Maybe @piscisaureus knows more than I do, either way we probably should verify this.

TL;DR: MSDN explains poorly how this works, and the default behavior (setting NumberOfConcurrentThreads to 0) is reasonable behavior.

Longer explanation: windows will not pre-empt a thread that's been woken up by `GetQueuedCompletionStatus` in order to wake up *another* thread blocked on the same completion port. However, once the the woken thread blocks on *something else*, then it is no longer treated as "associated with the completion port", and other threads will wake up. So practically speaking there is no problem; you can use as many threads as you like to poll a completion port, and they will all receive completions; the only consideration is that if threads do CPU-only work in response to a completion, the number of threads running concurrently will be limited to the number of CPUs in the system. 

Tangentially related: https://devblogs.microsoft.com/oldnewthing/20181123-00/?p=100305
